logins-api/android/LoginsApi/loginsapi/src/test/java/org/mozilla/sync15/logins/MemoryLoginsTest.kt

```diff
@@ -0,0 +1,199 @@
+/* Copyright 2018 Mozilla
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License. */
+package org.mozilla.sync15.logins
+
+import org.junit.Assert
+import org.junit.Test
+import org.junit.Assert.*
+import java.util.*
+import java.util.concurrent.CountDownLatch
+import java.util.concurrent.Semaphore
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicReference
+
+class MemoryLoginsTest {
+
+    fun getTestStore(): LoginsStorage {
+        return MemoryLoginsStorage(listOf(
+                ServerPassword(
+                        id = "aaaaaaaaaaaa",
+                        hostname = "https://www.example.com",
+                        httpRealm = "Something",
+                        formSubmitURL = null,
+                        username = "Foobar2000",
+                        password = "hunter2",
+                        timesUsed = 3,
+                        timeLastUsed = Date(2018, 7, 7).time,
+                        timeCreated = Date(2018, 7, 5).time,
+                        timePasswordChanged = Date(2018, 7, 6).time,
+                        usernameField = "users_name",
+                        passwordField = "users_password"
+                ),
+                ServerPassword(
+                        id = "bbbbbbbbbbbb",
+                        hostname = "https://www.example.org",
+                        httpRealm = null,
+                        formSubmitURL = "https://www.example.org/login",
+                        password = "MyVeryCoolPassword",
+                        timesUsed = 0,
+                        timeLastUsed = 0L,
+                        timeCreated = 0L,
+                        timePasswordChanged = 0L,
+                        username = null,
+                        usernameField = null,
+                        passwordField = null
+                )
+        ))
+    }
+    // TODO This should return T and not T?, I can't figure out how to coerce a T? to a T that
+    // doesn't throw if T is really nullable (e.g. `waitForResult<Int?>(SyncResult.fromValue(null))`
+    // should be fine!).
+    fun <T> waitForResult(e: SyncResult<T>): T? {
```

Comment:
If you really want this behaviour, use https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/BlockingQueue.html.  The real problem here isn't that you wait 2s for failure... it's that you wait 2s for _success_.

Reply:
I don't think I always wait two seconds for success. That's just the maximum time I wait. This is supposed to be equivalent to `Promise.race([new Promise(resolve => setTimeout(resolve, 2000)), result])`, and then waiting for that to resolve synchronously.

Reply:
(That said I do see how it might be cleaner with a 1 item blocking queue, and will try that)